Задание 1
Вернемся к набору данных о видеоиграх.

Ответьте на следующие вопросы:

1) Как критики относятся к спортивным играм?\
2) Критикам нравятся больше игры на PC или на PS4?\
3) Критикам больше нравятся стрелялки или стратегии?

Для каждого вопроса:

сформулируйте нулевую и альтернативную гипотезы;\
выберите пороговый уровень статистической значимости;\
опишите полученные результаты статистического теста.m

In [2]:
import pandas as pd
rat = pd.read_csv('C://Users//Ivan//Desktop//vgsales.csv')
rat.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  
0          51.0          8       322.0  Nintendo      E  
1           NaN        NaN         NaN       NaN    NaN  
2          73.0        8.3       709.0  Nintendo      E  
3          73.0          8       192.0  Nintendo      E  
4           NaN        NaN         NaN       NaN    NaN

<mark>1) Как критики относятся к спортивным играм?</mark>

1) H0: <u>средняя оценка критиками спортивных игр не отличается от оценок других жанров </u>\
   H1: <u>средняя оценка критиками спортивных игр различается от оценок других жанров </u>
   
2) Возьмем стандартный уровень статистической значимости в 0.05 

3) Используем двухвыборочный тест, т.к. две выборки

4) Используем двухсторонний вариант теста (и негативная и позитивная оценка) 

5) Определяем, равна ли дисперсия групп(критерий Уэлча).



In [5]:
# Создадим датасет только с оценками по спортивным играм и удаляем пустые строки
data_sport = rat[['Critic_Score']].loc[rat.Genre == 'Sports'].dropna()
# Создадим датасет только с оценками по играм остальных жанров и удаляем пустые строки
data_no_sport = rat[['Critic_Score']].loc[rat.Genre != 'Sports'].dropna()

In [7]:
import scipy.stats as stats
#Задаем уровень значимости
alpha = 0.05
#Применяем двухсторонний двухвыборочный критерий Стьюдента
result = stats.ttest_ind(data_sport['Critic_Score'], data_no_sport['Critic_Score'], equal_var=False)
print(result)
if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, средняя оценка критиками спортивных игр различается от оценок других жанров')
else:
    print('Не отвергаем нулевую гипотезу')

Ttest_indResult(statistic=8.08698828481822, pvalue=1.181171308320441e-15)
Отвергаем нулевую гипотезу, средняя оценка критиками спортивных игр различается от оценок других жанров


<mark>2. Критикам нравятся больше игры на PC или на PS4?</mark>

1) H0: <u>средняя оценка критиками игр на PC и PS4 одинакова </u>\
H1: <u>средняя оценка критиками игр на PC и PS4 различается </u>

2) Возьмем стандартный уровень статистической значимости в 0.05 

3) Используем двухвыборочный тест, т.к. две выборки

4) Используем двухсторонний вариант теста (т.к. нас интересует игры на какой платформе больше нравится критикам) 

5) Используем критерий Уэлча для определения равенства ли дисперсия групп

In [11]:
# Создадим датасет только с оценками критиков по играм на PC и удаляем пустые строки
data_PC = rat[['Critic_Score']].loc[rat.Platform == 'PC'].dropna()
# Создадим датасет только с оценками критиков по играм на PS4 и удаляем пустые строки
data_PS4 = rat[['Critic_Score']].loc[rat.Platform == 'PS4'].dropna()

In [12]:
#Задаем уровень значимости
alpha = 0.05
#Применяем односторонний двухвыборочный критерий Стьюдента
result = stats.ttest_ind(data_PC['Critic_Score'], data_PS4['Critic_Score'], equal_var=False)
print(result)
if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, средняя оценка критиками игр на PC различается от оценок игр на PS4')
else:
    print('Не отвергаем нулевую гипотезу')

Ttest_indResult(statistic=4.3087588262138725, pvalue=2.067249157283479e-05)
Отвергаем нулевую гипотезу, средняя оценка критиками игр на PC различается от оценок игр на PS4


Вывод: p_value < 0.05 -> отвергаем нулевую гипотезу о равенстве средних оценок критиков игр на PC и PS4, при этом средняя оценка на PC ниже средней оценки игр на PS4. С высокой долей статистической значимости можем предположить, что <b>критикам больше нравятся игры на PS4.</b>

<mark>3) Критикам больше нравятся стрелялки или стратегии?</mark>

1) H0: <u>средняя оценка критиками "стрелялок" и "стратегий" одинакова</u>\
H1: <u>средняя оценка критиками "стрелялок" и "стратегий" различается</u>
    
2)Возьмем стандартный уровень статистической значимости в 0.05

3)Используем двухвыборочный тест, т.к. две выборкит 

4)Используем двухсторонний вариант теста(т.к. нас интересует какой вид игры больше нравится критикам) 

5)Используем критерий Уэлча для определения равенства ли дисперсия групп

In [14]:
# Создадим датасет только с оценками критиков по играм на PC
data_Shooter = rat[['Critic_Score']].loc[rat.Genre == 'Shooter'].dropna()
# Создадим датасет только с оценками критиков по играм на PS4
data_Strategy = rat[['Critic_Score']].loc[rat.Genre == 'Strategy'].dropna()

In [15]:
#Задаем уровень значимости
alpha = 0.05
#Применяем односторонний двухвыборочный критерий Стьюдента
result = stats.ttest_ind(data_Shooter['Critic_Score'], data_Strategy['Critic_Score'], equal_var=False)
print(result)
if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, средняя оценка критиками "стрелялок" различается от оценок "стратегий"')
else:
    print('Не отвергаем нулевую гипотезу')

Ttest_indResult(statistic=-2.2972408230640315, pvalue=0.021938989522304823)
Отвергаем нулевую гипотезу, средняя оценка критиками "стрелялок" различается от оценок "стратегий"


Вывод: p_value > 0.05 -> не отвергаем нулевую гипотезу о равенстве средних оценок критиков игр на PC и PS4. С недостаточным уровнем статистической значимости можем сказать, что <b>средняя оценка стратегий выше средней оценки стрелялок.</b>

Реализуйте базовую модель логистической регрессии для классификации текстовых сообщений (используемые данные здесь) по признаку спама. 

Для этого:
1) Привидите весь текст к нижнему регистру;
2) Удалите мусорные символы;
3) Удалите стоп-слова;
4) Привидите все слова к нормальной форме;
5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)


Можете поэкспериментировать с параметрами TfidfVectorizer;
6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите random_state=42. Используйте train_test_split;
7) Постройте модель логистической регрессии, укажите random_state=42, оцените ее точность на тестовых данных;
8) Опишите результаты при помощи confusion_matrix;
9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [45]:

spam = pd.read_csv('https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/stat_case_study/spam.csv')
spam.head(5)

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [46]:
spam.describe()

Category                 Message
count      5572                    5572
unique        2                    5157
top         ham  Sorry, I'll call later
freq       4825                      30

In [47]:
import re
import nltk
nltk.download('stopwords')
nltk.download('wordnet')
from nltk.corpus import stopwords

from nltk.stem import WordNetLemmatizer



[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Ivan\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Ivan\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [48]:
def normalize_word(stroka):
    stroka = stroka.lower() #Приводим к нижнему регистру
    stroka = re.sub('[\W_]+',' ', stroka) # уберем все служебные символы
    stroka = stroka.split() # разделим все на слова
    stroka = [word for word in stroka if word not in set(stopwords.words('english'))] #убираем стоп-слова
    stroka = [wordnet_lemmatizer.lemmatize(word) for word in stroka] #приведем слова к леммам
    stroka = ' '.join(map(str, stroka)) #преобразуем в строку
    return stroka
spam['Message'] = spam.Message.apply(normalize_word)

In [49]:
print(spam.info(), spam.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Category  5572 non-null   object
 1   Message   5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB
None   Category                                            Message
0      ham  go jurong point crazy available bugis n great ...
1      ham                            ok lar joking wif u oni
2     spam  free entry 2 wkly comp win fa cup final tkts 2...
3      ham                u dun say early hor u c already say
4      ham                nah think go usf life around though


In [50]:
#5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)
tfidf_matrix.head()

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
1  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
3  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

   0125698789   02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
1         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
2         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
3         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
4         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   

    èn   ú1  〨ud  
0  0.0  0.0  0.0  
1  0.0  0.0  0.0  
2  0.0  0.0  0.0  
3  0.0  0.0  0.0  
4  0.0  0.0  0.0  

[5 rows x 8709 columns]

In [51]:
# Разделим данные на тестовые и тренировочные 30\70 
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(tfidf_matrix, spam['Category'], test_size=0.30, random_state=42)

In [52]:
# Построим модель логистической регресии, установим random_state=42
lr = LogisticRegression(random_state=42)

In [53]:
lr.fit(X_train, y_train) #Обучаем модель на тренировочных данных


LogisticRegression(random_state=42)

In [54]:
lr.predict(X_test) #Строим прогноз на тестовых данных


array(['ham', 'ham', 'ham', ..., 'ham', 'spam', 'ham'], dtype=object)

In [55]:
# Оценим ее точность на тестовых данных
from sklearn.metrics import accuracy_score
accuracy_score(y_test, lr.predict(X_test))

0.9635167464114832

<mark>что доля правильных ответов чуть более 0.95</mark>


In [58]:
# Опишем результаты при помощи confusion_matrix
from sklearn.metrics import confusion_matrix
cnf_matrix = confusion_matrix(y_test, lr.predict(X_test))
cnf_matrix

array([[1446,    2],
       [  59,  165]], dtype=int64)

Описание матрицы ошибок: True Positive: мы предсказали 'ham' верно в 1446 случаях.

True Negative: мы предсказали 'spam' в 165случаях

False Positive: мы предсказали что это 'ham', но на самом деле было 'spam' в 59 случаях.

False Negative: мы предсказали что это 'spam', но на самом деле было 'ham' в 2 случаях.

In [70]:
# Построим датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).
Original = pd.DataFrame(y_test) #Создаем DF с оригинальными (фактическими) тестовыми данными
Original.columns = ['original']
Original

original
3245      ham
944       ham
1044      ham
2484      ham
812       ham
...       ...
2505      ham
2525     spam
4975      ham
650      spam
4463      ham

[1672 rows x 1 columns]

In [71]:
Prediction = pd.DataFrame(lr.predict(X_test)) #Создаем DF с предсказанными данными
Prediction.columns = ['prediction']
Prediction

prediction
0           ham
1           ham
2           ham
3           ham
4           ham
...         ...
1667        ham
1668       spam
1669        ham
1670       spam
1671        ham

[1672 rows x 1 columns]

In [72]:
Result = Original.join(spam, how='left') #Джойним таблицу c оригинальными (фактическими) тестовыми данными к таблице с сообщениями 
Result

original Category                                            Message
3245      ham      ham  squeeeeeze christmas hug u lik frndshp den hug...
944       ham      ham  also sorta blown couple time recently id rathe...
1044      ham      ham  mmm thats better got roast b better drink 2 go...
2484      ham      ham                mm kanji dont eat anything heavy ok
812       ham      ham  ring come guy costume gift future yowifes hint...
...       ...      ...                                                ...
2505      ham      ham  hello boytoy made home constant thought love h...
2525     spam     spam  free entry 250 weekly comp send word win 80086...
4975      ham      ham  aiyo u poor thing u dun wan 2 eat u bathe already
650      spam     spam  1 000 cash 2 000 prize claim call09050000327 c...
4463      ham      ham  sorry flaked last night shit seriously goin ro...

[1672 rows x 3 columns]

In [73]:
Result = Result.drop(['Category'], axis=1) #Удаляем столбец с исходной категорией
Result = Result.reset_index(drop = True) #Сбрасываем индекс
Result.head()

original                                            Message
0      ham  squeeeeeze christmas hug u lik frndshp den hug...
1      ham  also sorta blown couple time recently id rathe...
2      ham  mmm thats better got roast b better drink 2 go...
3      ham                mm kanji dont eat anything heavy ok
4      ham  ring come guy costume gift future yowifes hint...

In [74]:
Result1 = pd.concat([Result, Prediction], axis = 1) #Делаем конкатенацию датафреймов
Result1

original                                            Message prediction
0         ham  squeeeeeze christmas hug u lik frndshp den hug...        ham
1         ham  also sorta blown couple time recently id rathe...        ham
2         ham  mmm thats better got roast b better drink 2 go...        ham
3         ham                mm kanji dont eat anything heavy ok        ham
4         ham  ring come guy costume gift future yowifes hint...        ham
...       ...                                                ...        ...
1667      ham  hello boytoy made home constant thought love h...        ham
1668     spam  free entry 250 weekly comp send word win 80086...       spam
1669      ham  aiyo u poor thing u dun wan 2 eat u bathe already        ham
1670     spam  1 000 cash 2 000 prize claim call09050000327 c...       spam
1671      ham  sorry flaked last night shit seriously goin ro...        ham

[1672 rows x 3 columns]

In [76]:
# Оставляем только строки отличающиеся предсказанным и исходным значением категории
Result1 = Result1[Result1.original != Result1.prediction]
Result1

original                                            Message prediction
40       spam  reminder downloaded content already paid goto ...        ham
47       spam  guess somebody know secretly fancy wanna find ...        ham
74       spam  oh god found number glad text back xafter msg ...        ham
84       spam  next amazing xxx picsfree1 video sent enjoy on...        ham
119      spam  rock yr chik get 100 filthy film xxx pic yr ph...        ham
...       ...                                                ...        ...
1509     spam  unique enough find 30th august www areyouuniqu...        ham
1525     spam  freemsg hi baby wow got new cam moby wanna c h...        ham
1567     spam    important customer service announcement premier        ham
1576      ham                                          free call       spam
1586     spam  sm ac blind date 4u rodds1 21 aberdeen united ...        ham

[61 rows x 3 columns]

In [77]:
61/1672

0.03648325358851675

В 3х % случаев неверно предсказали резутьтат

In [32]:
class Money():
    def __init__(self, dollars=0, cents=0):
        self.total_cents = 100 * dollars  + cents

    

    @property
    def dollars(self):
        return self.total_cents // 100

    @dollars.setter
    def dollars(self, value):
        if isinstance(value, int ) and value >= 0:
            self.total_cents = 100 * value + (self.total_cents % 100)
        else:
            print("Error dollars")

    @property
    def cents(self):
        return self.total_cents % 100

    @cents.setter
    def cents(self, value):
        if isinstance(value, int) and 0 <= value < 100:
            self.total_cents = 100*(self.total_cents // 100) + value
        else:
            print("Error cents")

    def __str__(self):
        print(f"Ваше состояние составляет {self.total_cents // 100} долларов {self.total_cents % 100} центов")
        

"""Создайте класс Money, у которого есть:
конструктор __init__, принимающий 2 аргумента: dollars, cents. 
По входным аргументам вам необходимо создать атрибут экземпляра total_cents. 
свойство геттер dollars, которое возвращает количество имеющихся долларов;
свойство сеттер dollars, которое принимает целое неотрицательное число - 
количество долларов и устанавливает при помощи него новое значение в атрибут экземпляра total_cents, 
при этом значение центов должно сохранятся. В случае, если в сеттер передано число, не удовлетворяющее условию, 
нужно печатать на экран сообщение "Error dollars";
свойство геттер cents, которое возвращает количество имеющихся центов;
свойство сеттер cents, которое принимает целое неотрицательное число меньшее 100 - количество центов и устанавливает 
при помощи него новое значение в атрибут экземпляра total_cents, при этом значение долларов должно сохранятся. 
В случае, если в сеттер передано число, не удовлетворяющее условию, нужно печатать на экран сообщение "Error cents";

метод __str__ (информация по данному методу), который возвращает строку вида "Ваше состояние составляет {dollars} 
долларов {cents} центов". Для нахождения долларов и центов в методе __str__ пользуйтесь свойствами
В экземпляр класса кроме атрибута total_cents сохранять ничего не нужно!"""

Bill = Money(101, 99)
#print(Bill)  # Ваше состояние составляет 101 долларов 99 центов
print(Bill.dollars, Bill.cents)  # 101 99
print(Bill.total_cents) # 10199


101 99
10199


In [33]:
Bill.dollars = 666
#print(Bill)  # Ваше состояние составляет 666 долларов 99 центов
Bill.cents = 12
Bill.dollars, Bill.cents
#print(Bill)  # Ваше состояние составляет 666 долларов 12 центов

(666, 12)

Задание 2
Реализуйте базовую модель логистической регрессии для классификации текстовых сообщений (используемые данные здесь) по признаку спама. Для этого:

1) Привидите весь текст к нижнему регистру;\
2) Удалите мусорные символы;\
3) Удалите стоп-слова;\
4) Привидите все слова к нормальной форме;\
5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:\

<b>from sklearn.feature_extraction.text import TfidfVectorizer</b>\

<b>tfidf = TfidfVectorizer()</b>\
<b>tfidf_matrix = tfidf.fit_transform(df.Message)</b>\
<b>names = tfidf.get_feature_names()</b>\
<b>tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)</b>\

Можете поэкспериментировать с параметрами TfidfVectorizer;\
6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите random_state=42. Используйте train_test_split;\
7) Постройте модель логистической регрессии, укажите random_state=42, оцените ее точность на тестовых данных;\
8) Опишите результаты при помощи confusion_matrix;\
9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).\